sdgym/benchmark.py

```diff
@@ -442,12 +443,130 @@ def _get_empty_dataframe(compute_quality_score, sdmetrics):
+                   Please make sure the path exists and permissions are given.""")
+    session = boto3.session.Session()
+    credentials = session.get_credentials()
+    synthesizer_string = 'synthesizers=['
+    for synthesizer in params['synthesizers']:
+        synthesizer_string += synthesizer.__name__ + ', '
+    synthesizer_string += ']'
+    # The indentation of the string is important for the python script
+    script_content = f"""import boto3
+from io import StringIO
+import sdgym
+from sdgym.synthesizers.sdv import (CopulaGANSynthesizer, CTGANSynthesizer, FastMLPreset,
+    GaussianCopulaSynthesizer, HMASynthesizer, PARSynthesizer, SDVRelationalSynthesizer,
+    SDVTabularSynthesizer,TVAESynthesizer)
+
+results = sdgym.benchmark_single_table(
+    {synthesizer_string}, custom_synthesizers={params['custom_synthesizers']},
+    sdv_datasets={params['sdv_datasets']},
+    additional_datasets_folder={params['additional_datasets_folder']},
+    limit_dataset_size={params['limit_dataset_size']},
+    compute_quality_score={params['compute_quality_score']},
+    sdmetrics={params['sdmetrics']}, timeout={params['timeout']},
+    detailed_results_folder={params['detailed_results_folder']},
+    multi_processing_config={params['multi_processing_config']}
+)
+
+# Convert DataFrame to CSV string
+csv_buffer = StringIO()
+results.to_csv(csv_buffer, index=False)
+s3 = boto3.client('s3',
+    aws_access_key_id='{credentials.access_key}',
+    aws_secret_access_key='{credentials.secret_key}')
+
+
+# Upload CSV to S3
+response = s3.put_object(
+    Bucket='{bucket_name}',
+    Key='{key_name}',
+    Body=csv_buffer.getvalue()
+)
+"""
+
+    return script_content
+
+
+def _create_instance_on_ec2(script_content):
+    ec2_client = boto3.client('ec2')
+    session = boto3.session.Session()
+    credentials = session.get_credentials()
+    print(f'This instance is being created in region: {session.region_name}') # noqa
+
+    # User data script to install the library
+    user_data_script = f"""#!/bin/bash
+    sudo apt update -y
+    sudo apt install python3-pip -y
+    echo "======== Install SDGYM ============"
+    sudo pip3 install sdgym
+    sudo pip3 install anyio
+    pip3 list
+    echo "======== Write Script ==========="
+    sudo touch ~/sdgym_script.py
+    echo "{script_content}" > ~/sdgym_script.py
+    echo "======== Run Script ==========="
+    sudo python3 ~/sdgym_script.py
+    echo "======== Terminate ==========="
+    sudo apt install awscli -y
+    aws configure set aws_access_key_id {credentials.access_key}
+    aws configure set aws_secret_access_key {credentials.secret_key}
+    aws configure set region {session.region_name}
+    INSTANCE_ID=$(curl -s http://169.254.169.254/latest/meta-data/instance-id)
+    aws ec2 terminate-instances --instance-ids $INSTANCE_ID
+    """
+
+    response = ec2_client.run_instances(
+        ImageId='ami-07d9b9ddc6cd8dd30',
+        InstanceType='t2.medium',
```

Comment:
Should these be configurable arguments on the function?

Reply:
This was one I was wondering about. This requires, the user to put in a lot more information about the configuration. The block mapping in line 548, those fields depend on the image. Instance type depends on the image. I think it's fine to hard code because these configurations exist and would require different scripts for each type of VM and errors on the ec2 instance (not enough memory, incorrect os script, etc.) are pretty much on the user to debug.

Reply:
Let's leave out configuration for now and add it later if we need to